Add support for metrics and native histograms

yzheqing opened the pull request:
This PR includes the basic features of Wavefront tier 2 Sender SDK, which includes:
1. wavefront metrics reporter
2. native wavefront histogram support
3. an abstraction that supports tagging at the host level

Testing is done through:
1. Unit tests for utility functions
2. End-to-end testing on the `tracing` server

It also includes an example file to test upon.

yzheqing committed:
Init npm package

yzheqing committed:
Add eslint and prettier formatter

yzheqing committed:
Implement basic structure of registry, reporter and wavefront histogram

yzheqing committed:
Create example for metric reporting

yzheqing committed:
Change to ES6 import style

yzheqing committed:
Add rollup and babel support

yzheqing committed:
update example

yzheqing committed:
Add comments, readme file, and a start section in package.json

github-gentleman363 commented on README.md:
```diff
@@ -0,0 +1,13 @@
+# wavefront-metrics-sdk
+This is a plugin for metrics which adds Wavefront reporters via direct ingestion) and an abstraction that supports tagging at the host level. 
```
Typo ")" after `direct ingestion`

github-gentleman363 commented on example.js:
```diff
@@ -0,0 +1,53 @@
+import TaggedRegistry from './src/registry';
+import WavefrontReporter from './src/wavefront-reporter';
+import { wavefrontHistogram } from './src/wavefrontHistogram';
```
You exported these in `src/index.js`. Didn't you mean to import from that file?

github-gentleman363 commented on src/util.js:
```diff
@@ -0,0 +1,40 @@
+
+function isEmpty(value) {
+  return (
+    (typeof value == 'string' && !value.trim()) ||
```
Do we need this line? L23 would handle the string case anyway.

yzheqing replied:
hmm there are other files using that function, could you specify which L23 it's referring to?

github-gentleman363 replied:
I see. Then, it makes more sense for the function name to be `isEmpty`. Sorry for the hassle.

github-gentleman363 commented on src/util.js:
```diff
@@ -0,0 +1,40 @@
+const tagSeparator = '-tags=';
+
+function isEmpty(value) {
```
I'd like to suggest a better name, either "isUndefOrNull" or "isNil" (this is a lodash term: https://lodash.com/docs/4.17.15#isNil).

github-gentleman363 commented on src/util.js:
```diff
@@ -0,0 +1,40 @@
+    return metricName;
+  }
+  let tagsType = Object.prototype.toString.call(tags).slice(8, -1);
+  if (tagsType === 'Object') {
```
To check if a variable type is "object", see https://stackoverflow.com/a/8511350/1885433

yzheqing replied:
thanks for pointing this out, but `null` has already been checked in `isNil(tags)` thus cannot be null if it enters this line

github-gentleman363 replied:
Yeah I agree with you on that but we can still do:
```suggestion
  if (typeof tags === "object") {
```

yzheqing replied:
oh...sorry for missing your point. I'll change the code accordingly.

github-gentleman363 commented on src/util.js:
```diff
@@ -0,0 +1,40 @@
+    let tagsArray = Object.keys(tags)
+      .sort()
+      .map(key => [key, tags[key]]);
+    return metricName + tagSeparator + JSON.stringify(tagsArray);
```
Use template strings: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

github-gentleman363 commented on src/wavefrontHistogram.js:
```diff
@@ -0,0 +1,88 @@
+import TaggedRegistry from './registry';
+import utils from './util';
+const metrics = require('metrics');
+const wavefrontSDK = require('wavefront-sdk-javascript');
```
Can we `import` these two?

yzheqing replied:
similar to this ticket [https://wavefront.atlassian.net/browse/MONIT-18946] when trying to import `metrics`; we can come back and revisit these two imports

github-gentleman363 commented on src/wavefrontHistogram.js:
```diff
@@ -0,0 +1,88 @@
+ * @return {WavefrontHistogram}
+ */
+function wavefrontHistogram(registry, name, tags = null) {
+  if (!name) {
```
Assuming you don't want `name` to be an empty string, you might also want to check if it's a string of blank spaces.
`if (!name || !name.trim())`

github-gentleman363 commented on src/wavefrontHistogram.js:
```diff
@@ -0,0 +1,88 @@
+    throw 'invalid counter name';
+  }
+
+  let histogram = new WavefrontHistogram();
```
Use `const`

github-gentleman363 commented on src/wavefrontHistogram.js:
```diff
@@ -0,0 +1,88 @@
+   * Instantiate brand new WavefrontHistogramImpl().
+   */
+  clear() {
+    this._delegate = wavefrontSDK.WavefrontHistogramImpl();
```
Don't you need `new`?

github-gentleman363 commented on src/wavefrontHistogram.js:
```diff
@@ -0,0 +1,88 @@
+  }
+}
+
+export { wavefrontHistogram, WavefrontHistogram };
```
Can we come up with more distinct names?

github-gentleman363 commented on src/wavefrontHistogram.js:
```diff
@@ -0,0 +1,88 @@
+  constructor(clockMillis = null) {
+    super();
+    this._delegate = new wavefrontSDK.WavefrontHistogramImpl(clockMillis);
+    this.update = value => this._delegate.update(value);
```
For consistency, we can define this function in the same manner other member functions are defined below.

yzheqing replied:
the `update` function is defined here because the `this` pointer cannot be preserved if not defined using the lambda

github-gentleman363 commented on src/registry.js:
```diff
@@ -0,0 +1,133 @@
+import utils from './util';
+const metrics = require('metrics');
```
Can we `import` this?

github-gentleman363 commented on src/registry.js:
```diff
@@ -0,0 +1,133 @@
+export default class TaggedRegistry extends Report {
+  constructor() {
+    super();
+  }
```
Don't need `constructor`

github-gentleman363 commented on src/registry.js:
```diff
@@ -0,0 +1,133 @@
+   * @returns {any | metrics.Meter | metrics.Timer | metrics.Counter | metrics.Histogram | metrics.Gauge|metrics.Metric}
+   */
+  _getOrAddMetric(metricName, tags, construct, supplier = null) {
+    let encodedName = utils.encodeKey(metricName, tags);
```
Use `const`

github-gentleman363 commented on src/registry.js:
```diff
@@ -0,0 +1,133 @@
+    let metric = this.getMetric(encodedName);
+    if (metric) return metric;
+
+    metric = supplier ? new construct(supplier) : new construct();
```
```suggestion
    metric = new construct();
```
Can be simplified.

yzheqing replied:
the `supplier` needs to be passed in when the `construct` is a gauge function

github-gentleman363 replied:
Oh sorry - I meant this:
```suggestion
    metric = new construct(supplier);
```

github-gentleman363 commented on src/registry.js:
```diff
@@ -0,0 +1,133 @@
+   * @returns {boolean}
+   */
+  hasCounter(key, tags = null) {
+    return this.getMetrics().counters.indexOf(utils.encodeKey(key, tags)) >= 0;
```
Are the return value of `this.getMetrics()` and `counters` guaranteed to be truthy always?
Otherwise, this line would throw a js error.

yzheqing replied:
yes the value is truthy. It'll be returned as an empty list if there is no `counter` under the `Reporter`

github-gentleman363 commented on src/wavefront-reporter.js:
```diff
@@ -0,0 +1,172 @@
+import { WavefrontHistogram } from './wavefrontHistogram';
+const metrics = require('metrics');
+const ScheduledReporter = metrics.ScheduledReporter;
+const wavefrontSDK = require('wavefront-sdk-javascript');
```
Can we `import` these?

github-gentleman363 commented on src/wavefront-reporter.js:
```diff
@@ -0,0 +1,172 @@
+   */
+  reportMinuteDistribution() {
+    this.histogramGranularity.add(wavefrontSDK.histogramGranularity.MINUTE);
+    return this;
```
Why return `this`?

yzheqing replied:
see example here: https://github.com/wavefrontHQ/wavefront-nodejs-metrics-sdk/blob/yzheqing/dev/example.js#L15
we might need the reference to the reporter after this function is called to set the granularity

github-gentleman363 replied:
hmm this style looks weird to me. If you were able to call this function, it means that you already had a reference to the class instance as well. In the example, you can do the following so that this function doesn't have to return `this`.
```
const wfReporter = new WavefrontReporter({
    server: server,
    token: token,
    source: 'wavefront-javascript-metrics',
    registry: reg,
    tags: { key1: 'val1', key2: 'val2' }
});
wfReporter.reportMinuteDistribution();
wfReporter.start(5000);
```

yzheqing replied:
This is something similar to a builder pattern. We need only one statement for creating an instance as well as setting its granularity. Furthermore, you could chain up like `new WavefrontReporter(...).reportMinuteDistribution(). reportHourDistribution().reportDayDistribution()`

github-gentleman363 commented on src/wavefront-reporter.js:
```diff
@@ -0,0 +1,172 @@
+  _decodeMetric(metric) {
+    const [metricName, metricTagsArray] = utils.decodeKey(metric.name);
+    const tagsArray = JSON.parse(metricTagsArray);
+    let metricTags = {};
```
Use `const`

github-gentleman363 commented on src/wavefront-reporter.js:
```diff
@@ -0,0 +1,172 @@
+          // If metric, client send_metric
+          for (let f of metricsFunctionMap[m.type]) {
+            const [suffix, value] = f(m);
+            console.log(metricName + suffix + ': ' + value);
```
Do we need this line?

github-gentleman363 commented on src/wavefront-reporter.js:
```diff
@@ -0,0 +1,172 @@
+          }
+        }
+      }
+    }
```
This whole `for` loop is not very easy to read. Can we refactor this?

yzheqing replied:
I left some comments to help understand the logic. Essentially it's just doing a loop through all the metrics and send to the wavefront server respectively based on its type.

yzheqing committed:
Resolve PR comments

github-gentleman363 commented on example.js:
```diff
@@ -1,6 +1,6 @@
-import { wavefrontHistogram } from './src/wavefrontHistogram';
+import { TaggedRegistry } from './src/index';
+import { WavefrontReporter } from './src/index';
+import { registerHistogram } from './src/index';
```
We can import these in the same line.

github-gentleman363 reviewed (commented):
@yzheqing I guess you're still working on addressing comments? I've found a couple of comments you haven't addressed yet.

yzheqing commented:
> @yzheqing I guess you're still working on addressing comments? I've found a couple of comments you haven't addressed yet.

Yes, I've been doing some end-to-end testing in the meantime. Thanks for your help in reviewing this repo and I've addressed all the comments for now.

github-gentleman363 commented on src/wavefrontHistogram.js:
```diff
@@ -34,14 +34,17 @@ class WavefrontHistogram extends metrics.Histogram {
-    this.update = value => this._delegate.update(value);
+  }
+
+  update(value, timestamp = null) {
```
`timestamp` is unused.

yzheqing replied:
This function is overriding its superclass's with a signature of `(value: number, timestamp?: number)`

github-gentleman363 reviewed (commented):
I left one comment and unresolved a couple of conversations that hadn't been addressed.

github-gentleman363 reviewed (commented):
LGTM
